hypershift-operator/controllers/nodepool/kubevirt/imagecacher.go

```diff
@@ -0,0 +1,248 @@
+		dv, err := qi.createDVForCache(ctx, cl, nodePool, infraId)
+		if err != nil {
+			return err
+		}
+		dvName = dv.Name
+	}
+
+	if len(dvName) == 0 {
+		return fmt.Errorf("can't get the name of the boot image cache DataVolume")
+	}
+	qi.dvName = dvName
+	qi.cleanOldCaches(ctx, cl, oldDVs)
+
+	return nil
+}
+
+func (qi *cachedQCOWImage) cleanOldCaches(ctx context.Context, cl client.Client, oldDVs []v1beta1.DataVolume) {
+	logger := ctrl.LoggerFrom(ctx)
+	for _, oldDV := range oldDVs {
+		if oldDV.DeletionTimestamp == nil {
+			logger.Info("deleting an old boot image cache DataVolume", "namespace", oldDV.Namespace, "DataVolume name", oldDV.Name)
+			err := cl.Delete(ctx, &oldDV)
+			if err != nil {
+				logger.Error(err, fmt.Sprintf("failed to delete an old DataVolume; namespace: %s, name: %s", oldDV.Namespace, oldDV.Name))
+			}
+		}
+	}
+}
+
+func (qi *cachedQCOWImage) createDVForCache(ctx context.Context, cl client.Client, nodePool *hyperv1.NodePool, infraId string) (*v1beta1.DataVolume, error) {
+	dv := qi.buildDVForCache(nodePool, infraId)
+
+	err := cl.Create(ctx, dv)
+	if err != nil && !errors.IsAlreadyExists(err) {
+		return nil, fmt.Errorf("failed to create a DataVolume for the boot image cache ; %w", err)
+	}
+
+	return dv, nil
+}
+
+func (qi *cachedQCOWImage) getDVSourceForVMTemplate() *v1beta1.DataVolumeSource {
+	return &v1beta1.DataVolumeSource{
+		PVC: &v1beta1.DataVolumeSourcePVC{
+			Namespace: qi.namespace,
+			Name:      qi.dvName,
+		},
+	}
+}
+
+func (qi *cachedQCOWImage) buildDVForCache(nodePool *hyperv1.NodePool, infraId string) *v1beta1.DataVolume {
+	var (
+		volumeSize   apiresource.Quantity
+		storageClass *string
+	)
+
+	kvPlatform := nodePool.Spec.Platform.Kubevirt
+
+	if kvPlatform.RootVolume != nil && kvPlatform.RootVolume.Persistent != nil {
+		if kvPlatform.RootVolume.Persistent.Size != nil {
+			volumeSize = *kvPlatform.RootVolume.Persistent.Size
+		}
+		storageClass = kvPlatform.RootVolume.Persistent.StorageClass
+	}
+
+	return &v1beta1.DataVolume{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: bootImageNamePrefix,
+			Namespace:    qi.namespace,
+			Labels: map[string]string{
+				bootImageDVLabelRoleName: bootImageDVLabelRoleValue,
+				bootImageDVLabelInfraId:  infraId,
+			},
+			Annotations: map[string]string{
+				bootImageDVAnnotationHash:          qi.hash,
+				cdiImmediateBindingAnnotation:      "true",
+				cdiDeleteAfterCompletionAnnotation: "false",
+			},
+		},
+		Spec: v1beta1.DataVolumeSpec{
+			Source: &v1beta1.DataVolumeSource{
+				HTTP: &v1beta1.DataVolumeSourceHTTP{
+					URL: qi.name,
+				},
+			},
+			Preallocation: pointer.Bool(true),
+			Storage: &v1beta1.StorageSpec{
+				AccessModes: []corev1.PersistentVolumeAccessMode{corev1.ReadWriteOnce},
+				Resources: corev1.ResourceRequirements{
+					Requests: map[corev1.ResourceName]apiresource.Quantity{
+						corev1.ResourceStorage: volumeSize,
+					},
+				},
+				StorageClassName: storageClass,
```

Comment:
storageclass and accessmodes are optional and shouldn't be set here unless the user defines them.

Reply:
@davidvossel - ack about the storage class. logically this is what you want because if the storage class is nil, then this is what we'll get here, but I'll make it more readable.

However I'm not 100% sure about the access mode. this DV is the cache DV itself - not the VM disk DV. I'm not sure it should be the same access mode. WDYT?